src/Simplex_tree/include/gudhi/Simplex_tree/Simplex_tree_iterators.h

```diff
@@ -455,6 +455,102 @@ class Simplex_tree_skeleton_simplex_iterator : public boost::iterator_facade<
   int curr_dim_;
 };
 
+/** \brief Iterator over the simplices of the simplicial complex that match the dimension specified by the parameter.
+ *
+ * Forward iterator, value_type is SimplexTree::Simplex_handle.*/
+template<class SimplexTree>
+class Simplex_tree_dimension_simplex_iterator : public boost::iterator_facade<
+    Simplex_tree_dimension_simplex_iterator<SimplexTree>,
+    typename SimplexTree::Simplex_handle const, boost::forward_traversal_tag> {
+ public:
+  typedef typename SimplexTree::Simplex_handle Simplex_handle;
+  typedef typename SimplexTree::Siblings Siblings;
+  typedef typename SimplexTree::Vertex_handle Vertex_handle;
+
+// any end() iterator
+  Simplex_tree_dimension_simplex_iterator()
+      : sib_(nullptr),
+        st_(nullptr),
+        dim_(0),
+        curr_dim_(0) {
+  }
+
+  Simplex_tree_dimension_simplex_iterator(SimplexTree const* st, int dim)
+      : sib_(nullptr),
+        st_(st),
+        dim_(dim),
+        curr_dim_(0) {
+    if (st == nullptr || st->root() == nullptr || st->root()->members().empty() || st->dimension() < dim) {
```

Comment:
```suggestion
    if (st == nullptr || st->root() == nullptr || st->root()->members().empty() || st->upper_bound_dimension() < dim) {
```
?
If we ask for the 0-skeleton, it feels wrong to iterate through the whole complex to compute its dimension before we start iterating on the vertices.
Ah, maybe the code that follows assumes that the range is not empty and misbehaves if it is, but then maybe that can be improved?

Should we also test `dim<0` here?

Is there a unit test for the empty complex?